src/extension.ts

```diff
@@ -20,51 +19,58 @@ import { activateLinters } from "./linters";
 import { activateFormatters } from "./formatters";
 import { SettingsKey, TaskQuickPickItem } from "./types";
 import { createLanguageServerClient } from "./lsp/common";
-import { dirname, relative } from "path";
+import { askShouldDownload, askConfigureOnOpen, askAndSelectRootDir, selectRootDir } from "./dialogs";
 
 export let extensionPath: string;
 export let workspaceState: vscode.Memento;
 let explorer: MesonProjectExplorer;
 let watcher: vscode.FileSystemWatcher;
 let compileCommandsWatcher: vscode.FileSystemWatcher;
-let mesonWatcher: vscode.FileSystemWatcher;
 let controller: vscode.TestController;
 
 export async function activate(ctx: vscode.ExtensionContext) {
   extensionPath = ctx.extensionPath;
   workspaceState = ctx.workspaceState;
 
-  if (!vscode.workspace.workspaceFolders) {
-    return;
-  }
-
-  const root = vscode.workspace.workspaceFolders[0].uri.fsPath;
-  const mesonFiles = await rootMesonFiles();
-  if (mesonFiles.length === 0) {
-    return;
-  }
-
-  let configurationChosen = false;
-  let savedMesonFile = workspaceState.get<string>("mesonbuild.mesonFile");
-  if (savedMesonFile) {
-    const filePaths = mesonFiles.map((file) => file.fsPath);
-    if (filePaths.includes(savedMesonFile)) {
-      configurationChosen = workspaceState.get<boolean>("mesonbuild.configurationChosen") ?? false;
+  // The workspace could contain multiple Meson projects. Take all root
+  // meson.build files we find. Usually that's just one at the root of the
+  // workspace.
+  const rootDirs = await mesonRootDirs();
+  let rootDir: string | undefined = undefined;
+  if (rootDirs.length == 1) {
+    rootDir = rootDirs[0];
+  } else if (rootDirs.length > 1) {
+    let savedSourceDir = workspaceState.get<string>("mesonbuild.sourceDir");
+    if (savedSourceDir && rootDirs.includes(savedSourceDir)) {
+      rootDir = savedSourceDir;
     } else {
-      savedMesonFile = undefined;
+      // We have more than one root meson.build file and none has been previously
+      // saved. Ask the user to pick one.
+      rootDir = await askAndSelectRootDir(rootDirs);
     }
   }
 
-  const mesonFile = savedMesonFile ?? mesonFiles[0].fsPath;
-  const sourceDir = dirname(mesonFile);
-  const buildDir = getBuildDirectory(sourceDir);
+  ctx.subscriptions.push(
+    vscode.commands.registerCommand("mesonbuild.selectRootDir", async () => {
+      let newRootDir = await selectRootDir(rootDirs);
+      if (newRootDir != rootDir) {
+        workspaceState.update("mesonbuild.sourceDir", newRootDir);
```

Comment:
This won't be saved reliably without `await`, because the `reloadWindow` command executed below will remove the workspace state object immediately.

(@tristan957 Please comment if this makes sense from the language point of view.)

Reply:
Makes sense, fixed.